### ISSUE TITLE
Sometimes first iOS key press acts like keyboard hide

### ISSUE THREAD

**Issue by Meligy** [ambistudio/react-native-spacer#3]
## Brief
When using the component as a wrapper for whole app, sometimes the first iOS key press ever acts like the keyboard is hidden. It runs the animation that hides stuff behind the keyboard.

Closing the keyboard later and opening it again would work just fine until I restart the app.

I did not try in Android.

Below I'll mention what seemed to fix it for me.

## Details

I was playing with the [sample app](https://github.com/nirsky/react-native-size-matters/tree/master/examples/expo-example-app) from react-native-size-matters. It had a chat window with textbox at the bottom that suffered from the keyboard issue.

After some messing up to upgrade the sample itself to react-native 0.55 (latest expo supports), and running it, and trying `KeyboardAvoidingView` then finding https://github.com/facebook/react-native/issues/13393#issuecomment-399722690, I modified the file that wraps all components:

https://github.com/nirsky/react-native-size-matters/blob/16383ea10bafcfd508cfbafb36995f4ea27b7f2a/examples/expo-example-app/components/behaviors/withScaledSheetSwitch.js#L9

Which originally looked like:

https://github.com/nirsky/react-native-size-matters/blob/16383ea10bafcfd508cfbafb36995f4ea27b7f2a/examples/expo-example-app/components/behaviors/withScaledSheetSwitch.js#L9-L17

        render() {
            return <View style={{flex: 1}}>
                <View style={{alignSelf: 'stretch', flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', margin: 10}}>
                    <Text style={{fontSize: 16}}>size-matters enabled?</Text>
                <Switch value={this.state.scale} onValueChange={() => this.setState(prevState => ({scale: !prevState.scale}))}/>
                </View>
                <ChildComponent {...this.props} scale={this.state.scale}/>
            </View>
        }

To now look like:

```
            return (
                <Spacer enabled style={{ height: '100%' }}>
                    <View  style={{flex:1, height: '100%'}}>
                            <View style={{ alignSelf: 'stretch', flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', margin: 10 }}>
                                <Text style={{ fontSize: 16 }}>size-matters enabled?</Text>
                                <Switch value={this.state.scale} onValueChange={() => this.setState(prevState => ({ scale: !prevState.scale }))} />
                            </View>
                            <ChildComponent {...this.props} scale={this.state.scale} />
                    </View>
                </Spacer>
            );
        }
```

(I only added the `<Spacer>` and styled it, and gave the `<View>` under it `height: '100%'`)

To imagine the app (pic from iPad, but I experienced the issue on most recent iPhone simulator, with iOS 12):

![image](https://user-images.githubusercontent.com/171637/46584411-8805c580-caae-11e8-8132-c39d4dd2523b.png)

Then when I opened the chat page in the app, I click on the textbox at the bottom, and the keyboard popped up with the screen shifting above it nicely, but when I clicked on the keyboard to enter text, the view expanded to full window, hiding the textbox behind the keyboard.

I clicked away, then clicked the textbox again, and tried typing. This time the textbox stayed higher than the keyboard. It worked just fine from then onwards. I reloaded the app and tried again, and oops, it failed the first time, but then it worked just fine the following times.

I repeated the test several times, disabled the scaling from `react-native-size-matters` to make sure it's not affecting anything, but same result. First key press on phone keyboard would expand the view behind the keyboard, but after hiding and showing the keyboard, pressing keys doesn't cause any harm.

## Investigation

I looked at the source code of the Spacer library, which was super awesome BTW, simplicity is great, and not easy, and I was able to fix the above problem by changing a few methods:

```
    _keyboardWillShow = ev => {
        if (this.props.enabled) {
            // Calculation new position above the keyboard
            let toValue = (this._locationY + this._viewHeight) - (windowHeight - (ev.endCoordinates.height + this._spaceMargin));
            this._animate(-1 * toValue).start();
        }
    };
```

```
    _setLayoutProps = ev => {
        this._container
                ._component
                .measureInWindow((x, y, width, height) => {
                    this._locationY = y;
                    this._viewHeight = height;
                })
    };
```

Both changes seemed to be needed. The first change in `_keyboardWillShow` removes the dependency on `this._container` in calculations. I'm not sure if this can be a problem sometimes (when the component is not outermost), which is part of the reason why I'm not creating a PR for this yet.

The other change felt more straightforward, and almost opposite the first one. It updates `this._viewHeight` based on the `this._container._component.measureInWindow()`.

## Extras

I also added a `componentWillUnmount`, but just because it felt right, not because of this issue.
It looked like:

```
    componentWillUnmount() {
        Keyboard.removeListener(showListenerEvent, this._keyboardWillShow);
        Keyboard.removeListener(hideListenerEvent, this._keyboardWillHide);
    }
```

**Comment by Meligy** [comment]
* Updated the post with correct source links
* Added a reference to https://github.com/facebook/react-native/issues/13393#issuecomment-399722690

**Comment by hieunc229** [comment]
Hi Meligy,

I had some trouble running with Expo on my laptop, though I managed added `Spacer` to `react-native-size-matters` and upload it [to an expo snack](https://snack.expo.io/@hieunc/react-native-spacer-and-size-matter) (please try it if you can).

I tried to replace `View` ([line 101 in expo-example-app/components/Chat.js](https://github.com/nirsky/react-native-size-matters/blob/16383ea10bafcfd508cfbafb36995f4ea27b7f2a/examples/expo-example-app/components/Chat.js#L101)) component with `Spacer` and it works fine on my iPhone. I think it would be the right place to use `Spacer`. (And I hope it was what you looking for).

Besides, in the function `_keyboardWillShow` of the current version, I don't recall exactly what happened, but I had run into an issue where the reference (`this._container`) was null and throw an undefined error, also `this._container._component` is not a standard property. So I added an extra calculation so it wouldn't be missed.

And thanks for the __Extras__, sound reasonable and pretty standard. I'll add it into the repo

**Comment by Meligy** [comment]
To reproduce the steps, I have pushed a working version of that repo with only updated expo (no spacer), which can be found here:
https://github.com/Meligy/react-native-size-matters/tree/expo30

I tried just installing the library and replacing line 101 in expo-example-app/components/Chat.js, taking away `View` and adding `Spacer` instead, and the issue still seemed there in the simulator (iOS 11 and 12).

Here's the full code with the minimal change:
https://github.com/Meligy/react-native-size-matters/tree/minimal-react-native-spacer/examples/expo-example-app
(Look at [Chat.js](https://github.com/Meligy/react-native-size-matters/blob/minimal-react-native-spacer/examples/expo-example-app/components/Chat.js))

It looked like this:
![2018-10-08_12-21-03](https://user-images.githubusercontent.com/171637/46589438-42202000-caf5-11e8-8365-2b26eff63ece.gif)

Only the first character seems to break it. Hiding and showing the keyboard seems just fine.

This is the patched version that seems to work for me:
https://github.com/Meligy/react-native-size-matters/tree/patched-react-native-spacer/examples/expo-example-app
(Look at [withScaledSheetSwitch.js](https://github.com/Meligy/react-native-size-matters/blob/patched-react-native-spacer/examples/expo-example-app/components/behaviors/withScaledSheetSwitch.js) and [spacer.js](https://github.com/Meligy/react-native-size-matters/blob/patched-react-native-spacer/examples/expo-example-app/components/behaviors/spacer.js))

**Comment by hieunc229** [comment]
Thanks for the gif, it super helpful. I will find out more about this issue!

**Comment by Meligy** [comment]
Thanks a lot.
I added some specific links to the files that were modified above.

**Comment by hieunc229** [comment]
I tried to remove `Spacer` then replace it with `View` component, and the issue isn't going away. So I suspect it was not the `Spacer` problem.

**Comment by Meligy** [comment]
With just a `View`, you never see the textbox when the keyboard shows up.

With the `Spacer`, the textbox is visible, except when you make the first keyboard character touch, it then disappears. Dismissing the keyboard and showing it again still shows the textbox correctly until you restart the app, and then the first keyboard character hides the textbox again.

**Comment by Meligy** [comment]
The patched version of spacer.js [here](https://github.com/Meligy/react-native-size-matters/blob/patched-react-native-spacer/examples/expo-example-app/components/behaviors/spacer.js) in [`patched-react-native-spacer`](https://github.com/Meligy/react-native-size-matters/blob/patched-react-native-spacer/examples/expo-example-app/) branch seems to work though (used in [withScaledSheetSwitch.js](https://github.com/Meligy/react-native-size-matters/blob/patched-react-native-spacer/examples/expo-example-app/components/behaviors/withScaledSheetSwitch.js)).

**Comment by hieunc229** [comment]
Can you verify that on Expo Snack, after reloading the app, and after the green sign "Device Connected" disappear, the issue still occur? I double checked when not using `Spacer`. It seems to me that somehow it rerender the view when the device reconnected.

Besides, I wouldn't recommend removing the extra calculation in case `this._component` return null (in the `_keyboardWillShow` function). It was there as because sometimes, for some reason, it actually returns null. 

Otherwise, feel free to create a folk and edit as you want.

**Comment by Meligy** [comment]
Hi again,
To remove any doubt about how Expo work, instead of doing React Native reload, just close the qpp and re-run expo again. Then it always happens again with the very first keyboard character press/tab.

For `_keyboardWillShow`, the calculation I removed was the one that depends on `_container`. It seems that even though `_container` was not falsy, the calculation was still causing the issue.

**Comment by hieunc229** [comment]
I can confirm that it was Expo somehow rerender components on the device after it connected (which doesn't make sense).

Here is a video shows the issue when I use `View` and try to tap on the chat box before "Device Connected". Also, when I switched to `Spacer` and wait for the device to completely connected, the problem disappears. [View on Youtube](https://youtu.be/If34X8_tNR0) (I realized the quality isn't that good)

Besides, on the `_keyboardWillShow`, the `this._component` is needed in case the child component position is not set to 'fixed'. Because once we scrolled the screen, the `Y position` of the child view changed (need to measure by the root view, not `window`), so it needs to pick up the new `Y position` in order to animate correctly. (My bad, I should have explained this earlier).

Again, feel free to folk and edit if you needed.
